pkg/sql/schemachanger/scbuild/internal/scbuildstmt/helpers.go

```diff
@@ -408,76 +408,116 @@ func indexColumnDirection(d tree.Direction) catpb.IndexColumn_Direction {
+	primary   *scpb.PrimaryIndex
+	secondary *scpb.SecondaryIndex
+	temporary *scpb.TemporaryIndex
+
+	name          *scpb.IndexName
+	partial       *scpb.SecondaryIndexPartial
+	partitioning  *scpb.IndexPartitioning
+	columns       []*scpb.IndexColumn
+	idxComment    *scpb.IndexComment
+	constrComment *scpb.ConstraintComment
 }
 
 // apply makes it possible to conveniently define build targets for all
-// the elements in the primaryIndexSpec.
-func (s primaryIndexSpec) apply(fn func(e scpb.Element)) {
-	fn(s.idx)
+// the elements in the indexSpec.
+func (s indexSpec) apply(fn func(e scpb.Element)) {
+	if s.primary != nil {
+		fn(s.primary)
+	}
+	if s.secondary != nil {
+		fn(s.secondary)
+	}
+	if s.temporary != nil {
+		fn(s.temporary)
+	}
 	if s.name != nil {
 		fn(s.name)
 	}
+	if s.partial != nil {
+		fn(s.partial)
+	}
 	if s.partitioning != nil {
 		fn(s.partitioning)
 	}
 	for _, ic := range s.columns {
 		fn(ic)
 	}
+	if s.idxComment != nil {
+		fn(s.idxComment)
+	}
+	if s.constrComment != nil {
+		fn(s.constrComment)
+	}
 }
 
-// clone conveniently deep-copies all the elements in the primaryIndexSpec.
-func (s primaryIndexSpec) clone() (c primaryIndexSpec) {
-	c.idx = protoutil.Clone(s.idx).(*scpb.PrimaryIndex)
+// clone conveniently deep-copies all the elements in the indexSpec.
+func (s indexSpec) clone() (c indexSpec) {
+	if s.primary != nil {
+		c.primary = protoutil.Clone(s.primary).(*scpb.PrimaryIndex)
+	}
+	if s.secondary != nil {
+		c.secondary = protoutil.Clone(s.secondary).(*scpb.SecondaryIndex)
+	}
+	if s.temporary != nil {
+		c.temporary = protoutil.Clone(s.temporary).(*scpb.TemporaryIndex)
+	}
 	if s.name != nil {
 		c.name = protoutil.Clone(s.name).(*scpb.IndexName)
 	}
+	if s.partial != nil {
+		c.partial = protoutil.Clone(s.partial).(*scpb.SecondaryIndexPartial)
+	}
 	if s.partitioning != nil {
 		c.partitioning = protoutil.Clone(s.partitioning).(*scpb.IndexPartitioning)
 	}
 	for _, ic := range s.columns {
 		c.columns = append(c.columns, protoutil.Clone(ic).(*scpb.IndexColumn))
 	}
+	if s.idxComment != nil {
+		c.idxComment = protoutil.Clone(s.idxComment).(*scpb.IndexComment)
+	}
+	if s.constrComment != nil {
+		c.constrComment = protoutil.Clone(s.constrComment).(*scpb.ConstraintComment)
+	}
 	return c
 }
 
-// makePrimaryIndexSpec constructs a primaryIndexSpec based on an existing
-// scpb.PrimaryIndex element.
-func makePrimaryIndexSpec(b BuildCtx, idx *scpb.PrimaryIndex) (s primaryIndexSpec) {
-	s.idx = idx
-	publicIdxTargets := b.QueryByID(idx.TableID).Filter(publicTargetFilter).Filter(hasIndexIDAttrFilter(idx.IndexID))
-	_, _, s.name = scpb.FindIndexName(publicIdxTargets)
-	_, _, s.partitioning = scpb.FindIndexPartitioning(publicIdxTargets)
-	scpb.ForEachIndexColumn(publicIdxTargets, func(_ scpb.Status, _ scpb.TargetStatus, ic *scpb.IndexColumn) {
+// makeIndexSpec constructs an indexSpec based on an existing index element.
+func makeIndexSpec(b BuildCtx, tableID catid.DescID, indexID catid.IndexID) (s indexSpec) {
```

Comment:
should this panic an error if it can't find the index? right now it'll just return an empty spec, no?

Reply:
Fixed.